lms/djangoapps/courseware/tests/test_views.py

```diff
@@ -308,6 +309,84 @@ def test_index_no_visible_section_in_chapter(self):
+        # create the _next parameter
+        courseware_url = "{courseware_url}?{active_block_id}".format(
+            courseware_url=reverse('courseware', kwargs={'course_id': unicode(course.id)}),
+            active_block_id=activate_block_id
+        )
+
+        # create the url for enroll_staff view
+        enroll_url = "{enroll_url}?next={courseware_url}".format(
+            enroll_url=reverse('enroll_staff', kwargs={'course_id': unicode(course.id)}),
+            courseware_url=courseware_url
+        )
+        return courseware_url, enroll_url
+
+    @ddt.data(
+        ({'enroll': "Enroll"}, 'courses/{}/courseware?activate_block_id={}', True),
+        ({'dont_enroll': "Don't enroll"}, '/courses/{}/about', False))
+    @ddt.unpack
+    def test_enroll_staff_redirection(self, data, expected_url_template, enrollment):
+        """
+        Verify unenrolled staff is redirected to correct url.
+        """
+        course = self._create_global_staff_user_and_course()
+        ___, enroll_url = self._create_url_for_enroll_staff(course)
+        response = self.client.post(enroll_url, data=data)
+
+        # Here we check the status code 302 because of the redirect
+        self.assertEqual(response.status_code, 302)
+        if enrollment:
+            self.assertRedirects(response, expected_url_template.format(unicode(course.id), unicode(course.location)))
+        else:
+            self.assertRedirects(response, expected_url_template.format(unicode(course.id)))
+
+    def test_enroll_staff_redirection_invalid_data(self):
+        """
+        If we try to post with an invalid data pattern, then we'll return a 404
+        """
+        course = self._create_global_staff_user_and_course()
+        ___, enroll_url = self._create_url_for_enroll_staff(course)
+        response = self.client.post(enroll_url, data={'test': "test"})
+        self.assertEqual(response.status_code, 302)  # todo goes to about page.
+
+    def test_courseware_redirection(self):
+        """
+        Test that when staff visiting courseware url, redirected to enroll_staff page.
+        post the valid data to enroll the staff in the course and verify it
+        staff has been enrolled to the given course.
+        """
+        course = self._create_global_staff_user_and_course()
+        courseware_url, enroll_url = self._create_url_for_enroll_staff(course)
+
+        response = self.client.get(courseware_url)
+        self.assertEqual(response.status_code, 302)
+        self.assertRedirects(response, enroll_url)
+
+        response = self.client.get(enroll_url)
+        self.assertEqual(response.status_code, 200)
+        response_content = response.content
+        self.assertIn('Enroll', response_content)
+        self.assertIn("dont_enroll", response_content)
+
+        response = self.client.post(enroll_url, data={'enroll': "Enroll"})
+        self.assertEqual(response.status_code, 302)
```

Comment:
it would be nice to know that we've ended up back on the `courseware_url`. There are a lot of redirects going on here, so you may find yourself using [`self.client.get(..., follow=True)`](https://docs.djangoproject.com/en/1.9/topics/testing/tools/#django.test.Client.get) to follow redirects.